### PR TITLE
Cosmos: Implement Binary Round-Trip Fuzzer

### LINKED ISSUE #4864: Cosmos: Operationalize the binary-encoding round-trip fuzzer (soak + scheduled run)
## Summary

The end-to-end **round-trip fuzzer** for Cosmos binary JSON encoding is implemented and merged as part of PR #4671, but it currently only runs on demand against a live account. This task tracks the work to **calibrate, schedule, and soak** it so it delivers the confidence it was designed for: validating that *millions of randomly-structured JSON documents* survive a full Cosmos round-trip unchanged across every binary-encoding configuration.

It directly closes the reviewer request (FabianMeiswinkel, PR #4671):

> a tool that can produce random json structures and does e2e validation (via canonicalization + hash) — this would allow us to test millions of differently structured json objects and increase confidence level.

## Background — what already exists

| Artifact | Location |
| -------- | -------- |
| Harness | `sdk/cosmos/azure_data_cosmos_perf/tests/binary_roundtrip_fuzzer.rs` |
| Design doc | `sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_ROUNDTRIP_FUZZER.md` |
| RFC (§7 round-trip invariant) | `sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_RFC.md` |

The harness already provides:

- a seeded (`SplitMix64`) generator of random JSON objects — bounded nesting, mixed/uniform arrays, arrays-of-objects, Unicode (BMP + astral) strings, and numbers within a calibrated envelope;
- a **Cosmos-compatible canonicalizer** (`canonicalize` / `canonicalize_number`) that models the backend's number-rewrite rules, so a faithful round-trip does not false-positive on `1.0 → 1`, `-0 → 0`, `u64`-above-`i64::MAX → f64`, etc.;
- three configs per document — **text control**, **binary**, and **binary + text-response** — so a mismatch localizes the broken layer (encoder / decoder / transcode);
- a **calibration mode** (`AZURE_COSMOS_FUZZ_CALIBRATE=true`) that prints a MATCH/DIFF table of `canonicalize_number` vs. the backend's returned form;
- deterministic reproduction: every run prints its seed; re-run with `AZURE_COSMOS_FUZZ_SEED=<seed>`.

The test is gated behind `--cfg test_category="binary_encoding"` and a live `AZURE_COSMOS_CONNECTION_STRING`, so it is **not** part of the default `cargo test` run or CI today.

## Goal / acceptance criteria

- [ ] Calibration re-run against the target live account(s) shows **all MATCH** (no DIFF) for the probe set; any new DIFF is modeled in `canonicalize_number` or the generator is narrowed, with the finding fed back into RFC §7.
- [ ] A **multi-day soak** (target: ≥ 1M documents, ideally 5M+) completes with zero unexplained mismatches. Seeds and config are recorded for reproduction.
- [ ] The fuzzer runs on a **schedule** (e.g. nightly/weekly pipeline against a dedicated test account), with failures surfacing the seed + both canonical forms for triage.
- [ ] Any real codec bug found is reduced to a minimal case, added as a **golden vector** (`binary_json/vectors.rs`) and/or an in-tree fuzz case, and fixed.
- [ ] (Stretch) Second account with dictionary encoding on/off wired via `AZURE_COSMOS_FUZZ_CONNECTION_STRING_2` so every doc runs through both.
- [ ] (Stretch) `--wide-numbers` coverage enabled once the wider numeric forms are calibrated.

## Scope of work

1. **Calibrate** number canonicalization against the account(s) the soak will target (calibration mode is a diagnostic — it prints, it does not assert).
2. **Provision** a dedicated fuzz account/database/container and wire the connection string into the pipeline secret store.
3. **Schedule** the soak run (release build) in a pipeline with a generous iteration budget and seed logging; capture and alert on failures.
4. **Triage loop**: on any mismatch, classify as (a) codec bug, (b) canonicalization gap, or (c) backend rewrite difference (escalate); add a regression vector for (a).
5. **Document** the run cadence and the triage runbook alongside the design doc.

## How to run

```bash
# Smoke run (a few hundred docs), local emulator:
AZURE_COSMOS_CONNECTION_STRING='AccountEndpoint=...;AccountKey=...;' \
AZURE_COSMOS_ALLOW_INVALID_CERT=true \
RUSTFLAGS='--cfg test_category="binary_encoding"' \
  cargo test -p azure_data_cosmos_perf --test binary_roundtrip_fuzzer -- --nocapture

# Multi-day soak (millions of docs), release build:
AZURE_COSMOS_CONNECTION_STRING='...' AZURE_COSMOS_FUZZ_ITERATIONS=5000000 \
AZURE_COSMOS_FUZZ_MAX_DEPTH=6 \
RUSTFLAGS='--cfg test_category="binary_encoding"' \
  cargo test -p azure_data_cosmos_perf --test binary_roundtrip_fuzzer --release -- --nocapture

# Calibrate number canonicalization (prints a MATCH/DIFF table, no assert):
AZURE_COSMOS_CONNECTION_STRING='...' AZURE_COSMOS_FUZZ_CALIBRATE=true \
RUSTFLAGS='--cfg test_category="binary_encoding"' \
  cargo test -p azure_data_cosmos_perf --test binary_roundtrip_fuzzer -- --nocapture

# Reproduce a failure exactly:
AZURE_COSMOS_FUZZ_SEED=<seed printed by the failing run> ... cargo test ...
```

### Environment knobs

| Variable | Default | Meaning |
| -------- | ------- | ------- |
| `AZURE_COSMOS_CONNECTION_STRING` | — (required) | live account (endpoint + key) |
| `AZURE_COSMOS_ALLOW_INVALID_CERT` | false | accept emulator cert |
| `AZURE_COSMOS_FUZZ_ITERATIONS` | 200 | number of generated docs |
| `AZURE_COSMOS_FUZZ_SEED` | random | PRNG seed (for reproduction) |
| `AZURE_COSMOS_FUZZ_MAX_DEPTH` | 5 | max JSON nesting depth |
| `AZURE_COSMOS_FUZZ_WIDE_NUMBERS` | false | widen numeric range (post-calibration) |
| `AZURE_COSMOS_FUZZ_UNICODE` | true | include Unicode strings |
| `AZURE_COSMOS_FUZZ_CALIBRATE` | false | number-calibration diagnostic mode |
| `AZURE_COSMOS_BINARY_TEST_DATABASE` / `_CONTAINER` | `binary-fuzz-*` | target names |

## What a failure means (triage)

1. **Real codec bug** — Rust encoded/decoded a value wrong. Reduce, add a golden vector + in-tree fuzz case, fix.
2. **Canonicalization gap** — the backend rewrote a number/string in a form `canonicalize_number` does not yet model. Tune it (design doc §3.1) or narrow the generator if the form is out of scope.
3. **Backend rewrite difference** — a genuinely different value after store; the highest-value finding — escalate.

## References

- PR #4671 — Cosmos binary JSON encoding
- Design doc: `BINARY_ENCODING_ROUNDTRIP_FUZZER.md`
- RFC §7 (round-trip invariant): `BINARY_ENCODING_RFC.md`

### PR DESCRIPTION
**Pull request by kundadebdatta**
## Summary

Adds **test/validation infrastructure** around the Cosmos binary-JSON codec (shipped in #4671). **No production code paths change.** Two complementary fuzzers plus golden-vector expansion, conformance tests, design docs, and CI wiring.

## What's included

**1. Live value-space round-trip fuzzer** — `azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs`

- Generates random JSON with a seeded PRNG, stores + reads each doc back through a **live account** under three encoding configs (text control, binary, binary+text-response), and asserts the value survives via a Cosmos-calibrated canonical form (number-normalize → RFC 8785 → SHA-256).
- Validates the **whole pipeline**: encode → wire → service store/rewrite → wire → decode.
- Every run prints its seed; reproduce a failure with `AZURE_COSMOS_FUZZ_SEED=<seed>`.
- Runs on the weekly **Live Test** stage (`binary_encoding` leg, pinned to ubuntu).

**2. Byte-level codec fuzzer (cargo-fuzz)** — `azure_data_cosmos_driver/fuzz/`

- libFuzzer targets feeding **arbitrary/mutated bytes** straight into the decoder (`decode` / `from_slice` / `transcode_to_text`) plus a decode→encode→decode differential — **no network**, so coverage guidance applies.
- Complements the deterministic in-tree `fuzz_tests.rs`; hardens the **protocol** itself (mis-encoded frames the encoder never produces).
- Runs on the weekly **Build** stage as a `-runs=0` golden-vector replay (Linux + nightly, non-blocking); coverage-guided soaks are manual/local.

## Round-trip fuzzer mechanism

```mermaid
sequenceDiagram
    participant H as Harness
    participant Enc as Binary codec
    participant Svc as Cosmos service
    H->>H: gen random JSON doc (seed)
    H->>H: sent_hash = canonical(normalize(doc))
    loop create / read / replace / upsert
        H->>Enc: encode(doc)
        Enc->>Svc: binary body (+ negotiation header)
        Svc-->>Enc: binary echo body
        Enc->>H: decode -> returned doc
        H->>H: assert canonical(returned) == sent_hash
    end
    Note over H: mismatch -> dump seed + both canonical forms
```

## Testing

- Offline unit tests (generator determinism, normalization, canonicalization) run on every PR.
- Live leg + weekly byte-fuzz replay are non-blocking (`ContinueOnError`).

### CODE REVIEW

File: sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs [hunk 1]
```diff
@@ -0,0 +1,2674 @@
...
+        "\u{0000}",
+        "😀",
+        "中文",
+        "\u{feff}",
+    ];
+    let s = if rng.below(2) == 0 {
+        EDGE[rng.below(EDGE.len() as u64) as usize].to_string()
+    } else {
+        gen_unicode_string(rng, cfg, 40)
+    };
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs:852]: Fixed. The fixed edge-case string pool is now ASCII-filtered when `unicode` is off (`edge.chars().filter(char::is_ascii)`), so `AZURE_COSMOS_FUZZ_UNICODE=false` runs no longer emit `😀`, `中文`, or the BOM.

File: sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs [hunk 2]
```diff
@@ -0,0 +1,2674 @@
...
+            // serialize→parse.
+            let (sent_canon, sent_hash) = canonical_hash(&normalize(&Value::Object(doc.clone())));
+
+            let context = format!("iter={iter} config={label} id={id} seed={}", cfg.seed);
+
+            // CREATE with content response (exercises the response decode path).
+            let created = with_transient_retry("create", &context, || {
+                container.create_item(&pk, &id, &doc, Some(write_options_with_content()))
+            })
+            .await?;
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs]: Fixed. The create path now has dedicated retry semantics: a `409 Conflict` seen on a *retried* attempt means a prior attempt already committed (the `(pk, id)` is unique per iteration+config), so the harness reads the stored item back instead of failing the soak. Genuine transient errors still back off and retry; non-transient errors still fail.

File: sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_ROUNDTRIP_FUZZER.md [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+> container framing (length/count prefixes, nested arrays-of-objects) and the
+> decoder's `MAX_DEPTH` guard. A `generator_depth_scales_with_max_depth` offline
+> test locks this in.
+
+### 4.1 Corpus-shaped documents
+
+Beyond the free-form hybrid documents, a configurable fraction of each run
+(`AZURE_COSMOS_FUZZ_SHAPE_RATIO`, default 50%) generates documents in the
+**shape of the real service testdata corpus**. A set of ~24 *shape samplers*
+(`SHAPE_SAMPLERS`) each reproduce the structure of one `testdata/*.json` family
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_ROUNDTRIP_FUZZER.md:288]: Fixed — the doc now says the default is 85%, matching `DEFAULT_SHAPE_RATIO`.

File: sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_ROUNDTRIP_FUZZER.md [hunk 4]
```diff
@@ -0,0 +1,629 @@
...
+
+| Config | `enabled` | `request_text_response` |
+| ------ | --------- | ----------------------- |
+| control (text) | false | — |
+| binary | true | false |
+| binary + text response | true | true |
+
+Extend with two accounts (dictionary encoding on/off) by pointing
+`AZURE_COSMOS_FUZZ_CONNECTION_STRING_2` at a second account — the harness runs
+every generated doc through both.
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_ROUNDTRIP_FUZZER.md]: Fixed. Removed the second-account claim; the harness reads a single `AZURE_COSMOS_CONNECTION_STRING`. Left a short note that the second-account (dictionary-encoding) mode is a planned extension, not a current feature.

File: sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs [hunk 5]
```diff
@@ -0,0 +1,2674 @@
...
+        }
+        Value::Array(items) => {
+            for item in items.iter_mut() {
+                bound_value(item, cfg);
+            }
+        }
+        Value::Object(map) => {
+            for v in map.values_mut() {
+                bound_value(v, cfg);
+            }
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs:1455]: Fixed. `bound_value`'s object arm now rebuilds the map, ASCII-filtering the **keys** as well as the values when `unicode` is off — property names go through the same binary string-encoding path, so this is required to actually isolate the ASCII codec path. Added a regression test (`unicode_off_strips_non_ascii_from_object_keys_and_values`).

File: sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs [hunk 6]
```diff
@@ -0,0 +1,2674 @@
...
+        for (label, client) in &clients {
+            let container = client
+                .database_client(&database_name)
+                .container_client(&container_name)
+                .await?;
+
+            // Distinct id per config so the three stores don't conflict.
+            let id = Uuid::new_v4().to_string();
+            let mut doc = base_doc.clone();
+            doc.insert("id".to_string(), Value::String(id.clone()));
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs]: Fixed. The item id is now derived deterministically from `(seed, iteration, config_index)` — `format!("fuzz-{:016x}-{iter}-{config_idx}", cfg.seed)` — without drawing from the document RNG stream. Rerunning with the printed `AZURE_COSMOS_FUZZ_SEED` now reproduces the exact document *and* its canonical form. (The `Uuid` in `run_calibration` is a throwaway storage key on a non-asserting diagnostic path and is intentionally left as-is.)

File: sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs [hunk 7]
```diff
@@ -0,0 +1,2674 @@
...
+        };
+        Self {
+            iterations: env_u64(ITERATIONS_ENV_VAR, DEFAULT_ITERATIONS),
+            seed,
+            max_depth: env_u64(MAX_DEPTH_ENV_VAR, DEFAULT_MAX_DEPTH as u64) as u32,
+            wide_numbers: env_bool(WIDE_NUMBERS_ENV_VAR, false),
+            unicode: env_bool(UNICODE_ENV_VAR, true),
+            breadth: env_u64(BREADTH_ENV_VAR, DEFAULT_BREADTH as u64).max(1) as u32,
+            shape_ratio: env_u64(SHAPE_RATIO_ENV_VAR, DEFAULT_SHAPE_RATIO as u64).min(100) as u32,
+            size_scale: env_u64(SIZE_SCALE_ENV_VAR, DEFAULT_SIZE_SCALE as u64).max(1) as u32,
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs]: Fixed. `breadth` is now clamped into `[1, u32::MAX]` *before* the `as u32` cast (`.clamp(1, u32::MAX as u64) as u32`), so a raw value that is a multiple of `2^32` (e.g. `4294967296`) no longer truncates to 0 and panics. Added a regression test (`breadth_env_clamps_multiple_of_2_pow_32_to_a_nonzero_u32`).

File: sdk/cosmos/azure_data_cosmos_driver/fuzz/README.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+# Minimize a crashing input:
+cargo +nightly fuzz tmin decode fuzz/artifacts/decode/crash-<hash>
+```
+
+## Thorough manual run on a Linux VM
+
+For deeper, longer fuzzing than the weekly CI budget (Option A caps each target
+at ~1000 s), run it by hand on any Linux box (or WSL2). This is the same thing CI
+does, without a wall-clock cap:
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos_driver/fuzz/README.md]: Fixed. Reworded to clarify that weekly CI only replays the committed corpus (`-runs=0`); coverage-guided mutation is a manual/local-only activity.

File: sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_ROUNDTRIP_FUZZER.md [hunk 9]
```diff
@@ -0,0 +1,629 @@
...
+RFC 8785 / `json-canon` refuses to serialize any integer at or beyond the JSON
+"max safe integer" (`2^53`), returning `Error("u64 must be less than JSON max
+safe integer")`. Cosmos, however, **preserves `i64` integers exactly** and stores
+`u64` above `i64::MAX` as lossy IEEE-754 doubles. To bridge this,
+`normalize_number` maps JCS-unsafe numbers to **stable string tokens** (an exact
+decimal for large `i64`, or the `f64` form for the lossy-double case), which are
+only ever compared for equality — never parsed back. This keeps the sent and
+round-tripped values comparable without tripping the JCS safe-integer guard, and
+is the concrete realization of the §9.2 "keep Cosmos number canonicalization"
+constraint. JCS-safe numbers still serialize as bare JSON numbers.
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_ROUNDTRIP_FUZZER.md:552]: Fixed. Updated the number-canonicalization section and the "key finding" note: Cosmos stores *every* JSON number as an IEEE-754 double, so integers beyond `2^53` (both `i64` and `u64`) are routed through `f64` and compared as the rounded double token — not preserved exactly. This now matches `normalize_number` and the recorded findings.

File: sdk/cosmos/eng/scripts/Run-BinaryJsonFuzz.ps1 [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+.DESCRIPTION
+    Installs (idempotently) the nightly toolchain + cargo-fuzz, seeds each
+    target's corpus from the committed golden vectors, then runs every fuzz
+    target under `azure_data_cosmos_driver/fuzz` for a bounded wall-clock budget.
+
+    This is a byte-level *protocol* fuzzer: it feeds arbitrary/mutated bytes
+    straight into the decoder (decode / from_slice / transcode_to_text) plus a
+    differential decode->encode->decode idempotence check. It is the complement
+    to the value-space live round-trip fuzzer in azure_data_cosmos_perf.
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/eng/scripts/Run-BinaryJsonFuzz.ps1]: Fixed — the description now points at `azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs`, where the live round-trip fuzzer actually lives.

File: sdk/cosmos/azure_data_cosmos_driver/fuzz/README.md [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+installs cargo-fuzz, seeds each corpus from the golden vectors, and runs every
+target under a wall-clock budget that scales by build reason:
+
+- **PR / CI smoke** — ~120 s/target (fast, non-blocking).
+- **Weekly deep run** — ~30 min/target on the scheduled build.
+
+Because it rides the already-registered `cosmos - ci` pipeline, the fuzz job
+shows up automatically as a job on every cosmos PR — no separate pipeline to
+register. Crash inputs are published as the `fuzz-crashes` build artifact so a
+failure can be reproduced and minimized (`cargo fuzz tmin`).
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos_driver/fuzz/README.md]: Fixed. Rewrote the section to describe the actual wiring: a weekly-only Build-stage `MatrixConfigs` entry that invokes `Run-BinaryJsonFuzz.ps1 -ValidateOnly` (libFuzzer `-runs=0` golden-corpus replay). Removed the `AdditionalMatrixConfigs` / per-PR / 120s-30min-soak claims.

File: sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_RFC.md [hunk 12]
```diff
@@ -0,0 +1,583 @@
...
+   value (`NumberUInt8` → `NumberInt16` → `NumberInt32` → `NumberInt64`; values
+   above `i64::MAX` use `NumberUInt64`). `[CROSS-VERIFY: .NET/C++]` — confirm the
+   .NET encoder prefers `Number*` over the extended `D7`–`DC` markers.
+3. **Floating-point (non-integer) numbers** → `NumberDouble` (`CC`).
+4. **Non-finite floats** → `null` (§4.3).
+5. **Strings** → system string if the value is in the system dictionary;
+   otherwise encoded-length (`< 64` bytes), `StrL1`, `StrL2`, or `StrL4` by
+   length. Compressed/base64/GUID/reference forms are **decoder-accepted but
+   encoder-optional** optimizations; a minimal conforming encoder need not emit
+   them. `[CROSS-VERIFY: .NET/C++]` — enumerate which optimizations the .NET
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_RFC.md]: Fixed as part of the §7 rewrite (see #7). The strings rule is now explicitly encoder-optional and notes that the Rust encoder never emits system/compressed strings — it uses encoded-length / `StrL*` forms. The "why deterministic encoding matters" note also clarifies that cross-SDK byte-equality only holds between encoders that made the same narrowing choices.

File: Cargo.toml [hunk 13]
```diff
@@ -117,6 +119,7 @@ hdrhistogram = "7.5"
 hostname = "0.4"
 hmac = { version = "0.12" }
 include-file = { version = "1.0.0", default-features = false }
+json-canon = "0.1"
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply Cargo.toml:122]: Fixed — added both `json_canon` and `json-canon`.

File: sdk/cosmos/azure_data_cosmos_driver/fuzz/Cargo.toml [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+version = "0.0.0"
+publish = false
+edition = "2021"
+license = "MIT"
+
+[package.metadata]
+cargo-fuzz = true
+
+[dependencies]
+libfuzzer-sys = "0.4"
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos_driver/fuzz/Cargo.toml:25]: Fixed — added both `libfuzzer_sys` and `libfuzzer-sys`.

File: sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_RFC.md [hunk 15]
```diff
@@ -0,0 +1,583 @@
...
+literal `05`, `NumberUInt8`, `Int8`, `Int16`, …; the string `"id"` can be a
+system string, an encoded-length string, `StrL1`, …). A **decoder MUST accept
+all valid encodings**. An **encoder MUST be deterministic**: for a given value it
+MUST emit exactly one, canonical encoding. The canonical rules are:
+
+1. **Integers in `[0,31]`** → literal small integer (single byte).
+2. **Other integers** → the narrowest fixed-width `Number*` marker that holds the
+   value (`NumberUInt8` → `NumberInt16` → `NumberInt32` → `NumberInt64`; values
+   above `i64::MAX` use `NumberUInt64`). `[CROSS-VERIFY: .NET/C++]` — confirm the
+   .NET encoder prefers `Number*` over the extended `D7`–`DC` markers.
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_RFC.md]: Fixed. Rewrote §7: an encoder MUST be *deterministic* but need not be *narrowest*. The Rust reference encoder deliberately emits a valid **verbose subset** (integers use `Int64`/`UInt64`, never the narrower `Number*` forms), and the rules are now marked "encoder-optional narrowing." This matches `conformance.rs`'s pinned snapshots (`encode_produces_expected_canonical_bytes`), so implementers following §7 no longer produce bytes that fail the Rust snapshots.

File: Cargo.toml [hunk 16]
```diff
@@ -88,6 +88,8 @@ version = "1.0.0"
 version = "1.0.0"
 
 [workspace.dependencies]
+arbitrary = { version = "1.4", features = ["derive"] }
+arbitrary-json = "0.1"
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply Cargo.toml:92]: Fixed — added `arbitrary`, `arbitrary_json`, and `arbitrary-json` to `eng/dict/crates.txt`.

File: sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_RFC.md [hunk 17]
```diff
@@ -0,0 +1,583 @@
...
+parse than UTF-8 JSON text. The service and every language SDK MUST agree on
+this format byte-for-byte.
+
+### 1.2 Scope
+
+This specification defines:
+
+- the byte-level layout of every value kind (§3–§6),
+- the **canonical** encoding a conforming encoder emits when multiple 
+  are valid (§7),
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_RFC.md:37]: Fixed — now reads "…when multiple **encodings** are valid (§7)".

File: sdk/cosmos/azure_data_cosmos/Cargo.toml [hunk 18]
```diff
@@ -144,6 +151,11 @@ name = "binary_encoding"
 path = "tests/binary_encoding.rs"
 required-features = ["key_auth", "control_plane", "fault_injection"]
 
+[[test]]
+name = "binary_roundtrip_fuzzer"
+path = "tests/binary_roundtrip_fuzzer.rs"
+required-features = ["key_auth", "fault_injection"]
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos/Cargo.toml]: Good catch. Added `control_plane` to the fuzzer test target's `required-features` and to every documented `--features` command (module docs, calibration rustdoc, and the design doc). The integration test now compiles with the documented invocation.

File: sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs [hunk 19]
```diff
@@ -0,0 +1,2791 @@
...
+
+/// Transient transport/service status codes worth retrying (429/408/503/... —
+/// includes the transport-generated 503 that carries DNS/connect blips on CI).
+/// Any other status (e.g. a decode failure) is treated as a genuine result.
+fn is_transient(err: &azure_data_cosmos::CosmosError) -> bool {
+    matches!(
+        u16::from(err.status().status_code()),
+        408 | 429 | 500 | 502 | 503 | 504
+    )
+}
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs:1775]: Good catch — this is legitimate and important. Binary response transcoding reports a broken decode as `HTTP 500 / SERIALIZATION_RESPONSE_BODY_INVALID` (sub-status 20020), and the status-only check treated every 500 as transient. A retry could then commit, the next attempt hit 409, and the conflict-recovery read masked the broken create-response decode path.

Fixed: `is_transient` now returns `false` when `sub_status() == SubStatusCode::SERIALIZATION_RESPONSE_BODY_INVALID`, so a genuine decode corruption surfaces immediately and is never retried into a masking conflict.

File: sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs [hunk 20]
```diff
@@ -0,0 +1,2791 @@
...
+            Value::Number(Number::from(i))
+        } else {
+            // i >= 2^53: the backend stores it as a lossy double, so tokenize the
+            // rounded f64 (not the exact decimal) — otherwise the returned,
+            // double-rounded value would mismatch.
+            Value::String(cosmos_double_token(i as f64))
+        }
+    } else if let Some(u) = n.as_u64() {
+        // u > i64::MAX: Cosmos stores it as a lossy double; token from the double.
+        Value::String(cosmos_double_token(u as f64))
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs]: Agreed. Representing a normalized wide number as a bare `Value::String` meant a codec bug that turned a wide number into the same decimal JSON string would canonicalize identically and pass silently.

Fixed: wide numbers now normalize to a distinct tagged value — `{ "$__cosmos_wide_number__": "<double token>" }` (`cosmos_wide_number_value` / `WIDE_NUMBER_TAG`) — so the token lives in its own type domain and can never collide with a genuine user string. The inner value stays a `String`, so `normalize_numbers` remains idempotent. Added a regression test (`wide_number_token_cannot_collide_with_a_plain_string`).

File: sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs [hunk 21]
```diff
@@ -0,0 +1,2791 @@
...
+                        .await
+                    {
+                        Ok(resp) => {
+                            break resp.into_model().map_err(|e| {
+                                format!("{context}: create response decode failed: {e}")
+                            })?;
+                        }
+                        Err(e)
+                            if e.status().status_code() == StatusCode::Conflict && attempt > 1 =>
+                        {
```
[the remark this answers is not in the document]
Comment (kundadebdatta) [reply sdk/cosmos/azure_data_cosmos/tests/binary_roundtrip_fuzzer.rs]: Correct — this is the interaction between the deterministic-id fix (#14) and the create retry. Because conflict recovery was gated on `attempt > 1`, a reproduced/completed seed hit `409` on the *first* create and failed before reaching the target document, defeating the advertised `AZURE_COSMOS_FUZZ_SEED` replay.

Fixed: the create path now recovers from `409 Conflict` on **any** attempt (reading the committed item back). This is safe precisely because of F1 — a broken create-response decode is `500 / 20020`, which is now non-transient and returns immediately, so it can never degrade into a conflict that the recovery read would mask.

File: sdk/cosmos/azure_data_cosmos/Cargo.toml [hunk 22]
```diff
@@ -52,6 +52,13 @@ azure_data_cosmos_driver = { path = "../azure_data_cosmos_driver", default-featu
 ] }
 azure_identity.workspace = true
 clap.workspace = true
+# Test-only deps for the binary round-trip fuzzer
+# (tests/binary_roundtrip_fuzzer.rs): it generates arbitrary JSON, canonicalizes
+# it (RFC 8785), and hashes the canonical form for differential comparison.
```
Comment (analogrelay) [inline sdk/cosmos/azure_data_cosmos/Cargo.toml:57]: Unnecessary comment IMO

File: sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_RFC.md
Comment (analogrelay) [inline sdk/cosmos/azure_data_cosmos_driver/docs/BINARY_ENCODING_RFC.md:1]: Can you separate this into a different PR? I'd like to develop a spec like this, but we should get the backend team engaged in reviewing it to make sure it's correct and not just hallucinated by Copilot :D

File: sdk/cosmos/azure_data_cosmos_driver/testdata/binary_json_vectors.json
Comment (analogrelay) [inline sdk/cosmos/azure_data_cosmos_driver/testdata/binary_json_vectors.json:1]: Before GA, I'd like us to run this corpus through a known-good implementation, like the .NET SDK (or even better if we could find a way to share it with the backend team and run it against whatever code they use as a codec). Because the reality is we can only hand-validate it so far :D

File: Cargo.toml [hunk 23]
```diff
@@ -88,6 +88,8 @@ version = "1.0.0"
 version = "1.0.0"
 
 [workspace.dependencies]
+arbitrary = { version = "1.4", features = ["derive"] }
```
Comment (heaths) [inline Cargo.toml:91]: Could you add a blurb to `/CONTRIBUTING.md` about fuzz testing and using `arbitrary` - even just a "Fuzzing" subheading under "Testing" and recommending use of `arbitrary` with a link to https://docs.rs/arbitrary would be fine. I want to avoid a situation where people keep adding similar but disparate dependencies for what should be shared. Or if you have more that you think other should use, just a brief write-up. Doesn't have to be extensive.

**Comment by kundadebdatta** [comment]
This PR is split into two:

- Part 1: [Round Trip Fuzzer](https://github.com/Azure/azure-sdk-for-rust/pull/4976)
- Part 2: [Binary Encoding RFC Fuzz](https://github.com/Azure/azure-sdk-for-rust/pull/4977).

Hence closing the original PR.
